### PR TITLE
feat: topology encryption utils and refactor

### PR DESCRIPTION
**Pull request by P1sar**
**ATTENTION THIS IS A BREAKING CHANGE**
1. I have added some CLI tools to make it easier to encrypt/test new topologies. So now we do not need to use weird online tools
2. I have added the AES encryption method with CTR mode. Pretty default.
3. Refactored AES decryption. So now it accepts []bytes instead of strings. Because Encryption should only encrypt bytes and not do any hex encoding/decoding
5. Refactored the way the encrypted topology map is parsed. The reasoning behind this is that the tool we used was forcing us to provide IV in string format while the body of the CipherText was in hex, this is really confusing. In reality, everything should be in the same format (eg Hex or base64, etc). Although it breaks old topology maps. 
6. So If this gets merged I will create a new issue to update topology on a testnet as well. 
7. Also in Many places Topology map usage were replaced to pointers since the size of the map could be pretty big

## Related Issue Or Context

Related to: https://github.com/sygmaprotocol/project-backlog/issues/60

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation

## Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have ensured that all acceptance criteria (or expected behavior) from issue are met
- [x] I have updated the documentation locally and in docs.
- [x] I have added tests to cover my changes.
- [ ] I have ensured that all the checks are passing and green, I've signed the CLA bot

### CODE REVIEW

File: docs/general/Topology.md [hunk 1]
```diff
@@ -25,4 +25,18 @@ To update the topology map, the map on the remote service needs to be updated. A
...
+IPFS should return hex formatted IV + data. To help you there are 2 utility CLI described below.
+
+## Utility CLI
+There are two utility CLI that helps encrypt and then test uploaded Toplogy.
+
+`./relayer topology encrypt --path ./topology.json --encryptionKey 123` 
+This command will encrypt provided topology and output correspponding hash and encrypted toplogy in hex representation (iv + data)
+
+`./relayer topology test --hash 123  --url https://cloudflare-ipfs.com/ipfs/123  --decryptionKey 321` 
+This command will fetch topology from IPFS and test it according to Relayers topology initialisation flow.
```
Comment (mpetrun5) [inline docs/general/Topology.md:40]:
```suggestion
This command will fetch topology from IPFS and test it according to Relayers topology initialization flow. 
```

File: topology/topology.go [hunk 2]
```diff
@@ -72,50 +74,60 @@ type TopologyProvider struct {
...
-	unecryptedBody := t.decrypter.Decrypt(string(body))
+	unecryptedBody := t.decrypter.Decrypt(ct)
 	rawTopology := &RawTopology{}
 	err = json.Unmarshal(unecryptedBody, rawTopology)
 	if err != nil {
-		return NetworkTopology{}, err
+		return nil, err
+	}
+	if hash != "" {
+		log.Info().Msgf("New NetworkTopology initialised. "+
```
Comment (mpetrun5) [inline topology/topology.go:108]: I would move this to app.go where it is intialized. Doesn't make sense to me to have if logic here to log something.

File: topology/encryption.go [hunk 3]
```diff
@@ -24,12 +26,27 @@ func NewAESEncryption(key []byte) (*AESEncryption, error) {
...
 
-func (ae *AESEncryption) Decrypt(data string) []byte {
-	iv := data[:aes.BlockSize]
-	bytes, _ := hex.DecodeString(data[aes.BlockSize:])
-
-	stream := cipher.NewCTR(ae.block, []byte(iv))
-	dst := make([]byte, len(bytes))
-	stream.XORKeyStream(dst, bytes)
+func (ae *AESEncryption) Decrypt(ct []byte) []byte {
+	iv := ct[:aes.BlockSize]
```
Comment (mpetrun5) [inline topology/encryption.go:30]: I'm fine with it all being hex. It was like that at some point, but checked out most tools and they accepted either as plaintext or as plaintext or hex and was more convenient that way.

File: topology/encryption.go [hunk 4]
```diff
@@ -24,12 +26,27 @@ func NewAESEncryption(key []byte) (*AESEncryption, error) {
...
 }
+
+// Encrypt is a function that encrypts provided bytes with AES in CTR mode
+// Returned value is iv + ct
+func (ae *AESEncryption) Encrypt(data []byte) []byte {
+	dst := make([]byte, len(data))
+	iv := make([]byte, 16)
+	_, err := rand.Read(iv)
+	if err != nil {
+		log.Fatalf("error while generating random string: %s", err)
```
Comment (mpetrun5) [inline topology/encryption.go:44]: This should just return error

**Review by mpetrun5 — CHANGES_REQUESTED** [review]

File: cli/topology/encryptTopology.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+var (
+	path          string
+	encryptionKey string
+)
+
+func init() {
+	encryptTopologyCMD.PersistentFlags().StringVar(&path, "path", "", "path to json file with network topology")
+	_ = encryptTopologyCMD.MarkFlagRequired("path")
+	encryptTopologyCMD.PersistentFlags().StringVar(&encryptionKey, "encryptionKey", "", "password to encrypt topology")
+	_ = encryptTopologyCMD.MarkFlagRequired("password")
```
[the remark this answers is not in the document]
Comment (P1sar) [reply cli/topology/encryptTopology.go:34]: It does not panic actually nor make it required. Weird. Fixed

[the remark this answers is not in the document]
Comment (mpetrun5) [reply cli/topology/encryptTopology.go:34]: It returns an error :smile:

File: topology/encryption.go (hunk 3 above)
Comment (P1sar) [reply topology/encryption.go:30]: I was actually trying to find some RFC that can describe an implementation of AES CTR mode to be sure In my thoughts. Generally speaking, IV is not a part of the AES standard (AES is a permutation network first of all). IV is a part of the Modes of Operation standard and appears there is a lot of confusion here. 
But I remember for sure that I always was using everything in hex.

File: docs/general/Topology.md [hunk 6]
```diff
@@ -25,4 +25,17 @@ To update the topology map, the map on the remote service needs to be updated. A
...
+IPFS should return hex formatted IV + data. To help you there are 2 utility CLI described below.
+
+## Utility CLI
+
+`./relayer topology encrypt --path ./topology.json --encryptionKey 123` 
+This command will encrypt provided topology and output corresponding hash and encrypted toplogy in hex representation (iv + data)
+
+`./relayer topology test --hash 123  --url https://cloudflare-ipfs.com/ipfs/123  --decryptionKey 321` 
+This command will fetch topology from IPFS and test it according to Relayers topology initialization flow. 
+This allows to test correctnes of topology before actually call `RefreshKey`
```
Comment (mpetrun5) [inline docs/general/Topology.md:40]:
```suggestion
This allows to test correctnes of topology before actually calling `RefreshKey`
```

**Review by mpetrun5 — APPROVED** [review]

**Review by MakMuftic — APPROVED** [review]

**Pull request merged**